pyccel/parser/semantic.py

```diff
@@ -1570,6 +1570,14 @@ def _assign_lhs_variable(self, lhs, d_var, rhs, new_expressions, is_augassign =
                                         shape=a.alloc_shape, status=status))
                             args = new_args
                             new_args = []
+                    elif isinstance(lhs.class_type, (HomogeneousListType, HomogeneousSetType,DictType)):
+                        if isinstance(rhs, (PythonList, PythonDict, PythonSet)):
+                            alloc_type = 'init'
+                        elif rhs.get_attribute_nodes(IndexedElement):
+                            alloc_type = 'resize'
+                        else:
+                            alloc_type = 'reserve'
```

Comment:
I don't understand the logic behind this. Can you explain?

Reply:
This is to differentiate between the uses of `resize` and `reserve`; `resize` actually sets the size; and thus the capacity; of the container to the desired size whereas `reserve` only sets its capacity and keeps its size to 0. The former should be used when we wanna access index elements of the container directly while the latter can be used when append/add will be used afterward and we only wanna reserve the adequate size to avoid repetitive reallocations that may stem from size doubling. Therefore, I need to check if the code block in the `rhs` has any `IndexedElement` nodes. My assumption that their existence is tied to them being used on the `lhs` variable might not always be true. Shall I double check this as well?

Reply:
So the first case handles:
```python
def f(a : list[int]):
    b = a[1:-1]
```
while the second case handles everything else?

Yes it would be good to check this and maybe add this case in the valgrind tests. Could you also add a comment about this here.

Reply:
Sure thing!
by the way, I could handle the issue we have with `b = a` here as well. Shall I do it here or does it need a PR of its own?

Reply:
Which issue?

Reply:
```python

if __name__ == "__main__":
    a = [1,2,3]
    b = a
    a[0] = 2
    print(b[0])
```
this code prints 1 instead of 2.
`b` should only be a pointer and not a new list

Reply:
That doesn't sound related to this issue to me so I think we should fix it separately. Can you create an issue?

Reply:
It already exists #2008

Reply:
Ah super :+1: